transformer_lens/hook_points.py

```diff
@@ -18,45 +27,51 @@
 class HookPoint(nn.Module):
     def __init__(self):
         super().__init__()
-        self.fwd_hooks = []
-        self.bwd_hooks = []
+        self.fwd_hooks: List[LensHandle] = []
+        self.bwd_hooks: List[LensHandle] = []
         self.ctx = {}
 
         # A variable giving the hook's name (from the perspective of the root
         # module) - this is set by the root module at setup.
         self.name = None
 
-    def add_hook(self, hook, dir="fwd"):
+    def add_perma_hook(self, hook, dir="fwd") -> None:
+        self.add_hook(hook, dir=dir, perma=True)
+
+    def add_hook(self, hook, dir="fwd", is_permanent=False) -> None:
         # Hook format is fn(activation, hook_name)
         # Change it into PyTorch hook format (this includes input and output,
         # which are the same for a HookPoint)
-
         if dir == "fwd":
 
             def full_hook(module, module_input, module_output):
                 return hook(module_output, hook=self)
 
             handle = self.register_forward_hook(full_hook)
+            handle = LensHandle(handle, is_permanent)
             self.fwd_hooks.append(handle)
         elif dir == "bwd":
             # For a backwards hook, module_output is a tuple of (grad,) - I don't know why.
             def full_hook(module, module_input, module_output):
                 return hook(module_output[0], hook=self)
-
             handle = self.register_full_backward_hook(full_hook)
+            handle = LensHandle(handle, is_permanent)
             self.bwd_hooks.append(handle)
         else:
             raise ValueError(f"Invalid direction {dir}")
 
-    def remove_hooks(self, dir="fwd"):
-        if (dir == "fwd") or (dir == "both"):
-            for hook in self.fwd_hooks:
-                hook.remove()
-            self.fwd_hooks = []
-        if (dir == "bwd") or (dir == "both"):
-            for hook in self.bwd_hooks:
-                hook.remove()
-            self.bwd_hooks = []
+    def remove_hooks(self, dir="fwd", including_permanent=False) -> None:
+
+        def _remove_hooks(hooks: list[LensHandle]) -> None:
+            for handle in hooks:
+                if handle.is_permanent or including_permanent:
+                    handle.hook.remove()
+            return [] if including_permanent else [hook for hook in hooks if hook.is_permanent]
```

Comment:
Seems more elegant to create an `output_hooks=[]` at the start of the function, and add an else statement to the for loop to add to the output list? More robust to bugs, and to changes to the function.